### PR TITLE
BUG: Retain tz-aware dtypes with melt (#15785)

### LINKED ISSUE #15785: BUG: melt changes type of tz-aware columns
#### Code Samples

```python
import pandas as pd
frame = pd.DataFrame({'klass':range(5), 'ts': [pd.Timestamp('2017-03-23 08:22:42.173378+01'), pd.Timestamp('2017-03-23 08:22:42.178578+01'), pd.Timestamp('2017-03-23 08:22:42.173578+01'), pd.Timestamp('2017-03-23 08:22:42.178378+01'), pd.Timestamp('2017-03-23 08:22:42.163378+01')], 'attribute':['att1', 'att2', 'att3', 'att4', 'att5'], 'value': ['a', 'b', 'c', 'd', 'd']})
# At this point, frame.ts is of dtype datetime64[ns, pytz.FixedOffset(60)]
frame.set_index(['ts', 'klass'], inplace=True)
queried_index = frame.query('value=="d"').index
pivoted_frame = frame.reset_index().pivot_table(index=['klass', 'ts'], columns='attribute', values='value', aggfunc='first')
melted_frame = pd.melt(pivoted_frame.reset_index(), id_vars=['klass', 'ts'], var_name='attribute', value_name='value')
# At this point, melted_frame.ts is of dtype datetime64[ns]
queried_after_melted_index = melted_frame.query('value=="d"').set_index(['ts', 'klass']).index
frame.loc[queried_index]  # Works
frame.loc[queried_index] = 'test'  # Works
frame.loc[queried_after_melted_index]  # Works
frame.loc[queried_after_melted_index] = 'test'  # Breaks
```
The last statement gives: 
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.5/dist-packages/pandas/core/indexing.py", line 140, in __setitem__
    indexer = self._get_setitem_indexer(key)
  File "/usr/local/lib/python3.5/dist-packages/pandas/core/indexing.py", line 127, in _get_setitem_indexer
    return self._convert_to_indexer(key, is_setter=True)
  File "/usr/local/lib/python3.5/dist-packages/pandas/core/indexing.py", line 1230, in _convert_to_indexer
    raise KeyError('%s not in index' % objarr[mask])
KeyError: "MultiIndex(levels=[[2017-03-23 07:22:42.163378, 2017-03-23 07:22:42.173378, 2017-03-23 07:22:42.173578, 2017-03-23 07:22:42.178378, 2017-03-23 07:22:42.178578], [0, 1, 2, 3, 4]],\n           labels=[[3, 0], [3, 4]],\n           names=['ts', 'klass']) not in index"
```

#### Problem description

- It is counter-intuitive that any operation (which does not explicitly mention in its docs that it does) alters the type of any column.
- Also counter-intuitive is that ```frame.loc``` has different behavior in a statement than it has in an assignment.

#### Expected Output
- ```melted_frame.ts``` and ```frame.ts``` have the same dtype.
- ```DataFrame.loc``` fails in both cases, not just in an assignment, or succeeds in both.

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.5.2.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.0-66-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.19.2
nose: None
pip: 9.0.1
setuptools: 20.7.0
Cython: None
numpy: 1.12.0
scipy: None
statsmodels: None
xarray: None
IPython: 5.3.0
sphinx: None
patsy: None
dateutil: 2.6.0
pytz: 2016.10
blosc: None
bottleneck: None
tables: None
numexpr: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: 0.7.3
lxml: 3.5.0
bs4: 4.4.1
html5lib: 0.999
httplib2: 0.9.1
apiclient: None
sqlalchemy: None
pymysql: None
psycopg2: 2.6.1 (dt dec pq3 ext lo64)
jinja2: 2.8
boto: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] closes #15785
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

`.values` call was converting tz aware data to tz naive data (by casting to a numpy array). Added an additional test for Categorical data as well.

### CODE REVIEW

File: pandas/core/reshape/melt.py [hunk 1]
```diff
@@ -70,7 +72,13 @@ def melt(frame, id_vars=None, value_vars=None, var_name=None,
 
     mdata = {}
     for col in id_vars:
-        mdata[col] = np.tile(frame.pop(col).values, K)
+        id_data = frame.pop(col)
+        if is_extension_type(id_data):
+            # Preserve pandas dtype by not converting to a numpy array
+            id_data = concat([id_data] * K, ignore_index=True)
```
Comment (jreback) [inline pandas/core/reshape/melt.py]: @TomAugspurger do we have this in ExtensionArray ATM? ``.tile()``? or could emulate like this

File: pandas/core/reshape/melt.py [hunk 2]
```diff
@@ -70,7 +72,13 @@ def melt(frame, id_vars=None, value_vars=None, var_name=None,
 
     mdata = {}
     for col in id_vars:
-        mdata[col] = np.tile(frame.pop(col).values, K)
+        id_data = frame.pop(col)
+        if is_extension_type(id_data):
+            # Preserve pandas dtype by not converting to a numpy array
```
Comment (jreback) [inline pandas/core/reshape/melt.py]: comment is not needed

File: pandas/tests/reshape/test_melt.py [hunk 3]
```diff
@@ -212,6 +212,33 @@ def test_multiindex(self):
...
+        pd.Series(pd.date_range('2010', periods=5, tz='US/Pacific')),
+        pd.Series(["a", "b", "c", "a", "d"], dtype="category")])
+    @pytest.mark.parametrize("pandas_dtype_value", [True, False])
+    def test_pandas_dtypes_id_var(self, col, pandas_dtype_value):
+        # GH 15785
+        # Pandas dtype in the id
+        df = DataFrame({'klass': range(5),
+                        'col': col,
+                        'attr1': [1, 0, 0, 0, 0]})
+        if pandas_dtype_value:
```
Comment (jreback) [inline pandas/tests/reshape/test_melt.py]: can you move what this does into the parameterize, maybe by using a fixture. this defeats the purpose of being able to look at the parameterize and see what cases are being tested

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/reshape/test_melt.py [hunk 4]
```diff
@@ -212,6 +212,38 @@ def test_multiindex(self):
         res = self.df1.melt()
         assert res.columns.tolist() == ['CAP', 'low', 'value']
 
+    @pytest.mark.parametrize("col", [
+        pd.Series(pd.date_range('2010', periods=5, tz='US/Pacific')),
+        pd.Series(["a", "b", "c", "a", "d"], dtype="category")])
+    def test_pandas_dtypes_id_var(self, col):
```
[the remark this answers is not in the document]
Comment (jreback) [reply pandas/tests/reshape/test_melt.py]: ahh I see @WillAyd suggested. in any event this makes this much harder to read.

File: pandas/core/reshape/melt.py (hunk 1 above)
Comment (TomAugspurger) [reply pandas/core/reshape/melt.py]: No `.tile`. We do have `_concat_same_type`.

File: doc/source/whatsnew/v0.23.0.txt [hunk 5]
```diff
@@ -896,6 +896,7 @@ Timezones
 - Bug in :func:`Timestamp.tz_localize` where localizing a timestamp near the minimum or maximum valid values could overflow and return a timestamp with an incorrect nanosecond value (:issue:`12677`)
 - Bug when iterating over :class:`DatetimeIndex` that was localized with fixed timezone offset that rounded nanosecond precision to microseconds (:issue:`19603`)
 - Bug in :func:`DataFrame.diff` that raised an ``IndexError`` with tz-aware values (:issue:`18578`)
+- Bug in :func:`melt` that coverted tz-aware dtypes to tz-naive (:issue:`15785`)
```
Comment (jschendel) [inline doc/source/whatsnew/v0.23.0.txt]: coverted --> converted

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!

**Pull request merged**